src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,567 @@
+
+        mov             rdi, r12
+        call            C_FUNC(ReverseLeaveRuntimeHelper)
+        mov             byte ptr [rbp + UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET], 0
+
+        // restore rax, xmm0
+        mov             rax, [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h]
+        movdqa          xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h]
+        
+        jmp             LOCAL_LABEL(Done_NotifyHost_ReverseLeaveRuntime)
+#endif
+
+LOCAL_LABEL(WrongAppDomain):
+        //
+        // home register args to the stack
+        //
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h], rdi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h], rsi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h], rdx
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 18h], rcx
+
+        //
+        // save off xmm registers
+        //
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h], xmm0
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 10h], xmm1
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 20h], xmm2
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h], xmm3
+
+        //
+        // call our helper to perform the AD transtion 
+        //
+        mov             rdi, METHODDESC_REGISTER
+        lea             r8,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET]
+        mov             rax, [METHODDESC_REGISTER + OFFSETOF__UMEntryThunk__m_pUMThunkMarshInfo]
+        mov             r9d, [rax + OFFSETOF__UMThunkMarshInfo__m_cbActualArgSize]
+        call            C_FUNC(UM2MDoADCallBack)
+
+        // restore return value
+        mov             rax,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h]
+        movdqa          xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h]
+
+        jmp             LOCAL_LABEL(PostCall)
+
+NESTED_END UMThunkStub, _TEXT
+
+//
+// EXTERN_C void __stdcall UM2MThunk_WrapperHelper(
+//       void *pThunkArgs,               // rdi
+//       int argLen,                     // rsi
+//       void *pAddr,                    // rdx            // not used
+//       UMEntryThunk *pEntryThunk,      // rcx
+//       Thread *pThread)//               // [entry_sp + 28h]
+//
+NESTED_ENTRY UM2MThunk_WrapperHelper, _TEXT, NoHandler
+
+// number of integer registers saved in prologue
+#define UM2MThunk_WrapperHelper_NUM_REG_PUSHES 3
+
+#define UM2MThunk_WrapperHelper_CALLEE_SCRATCH_SIZE SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES
+#define UM2MThunk_WrapperHelper_STACK_FRAME_SIZE 0x38
+
+#define UM2MThunk_WrapperHelper_FRAME_OFFSET UM2MThunk_WrapperHelper_CALLEE_SCRATCH_SIZE
+#define UM2MThunk_WrapperHelper_FIXED_STACK_ALLOC_SIZE (UM2MThunk_WrapperHelper_STACK_FRAME_SIZE - (UM2MThunk_WrapperHelper_NUM_REG_PUSHES * 8))
+
+        push_nonvol_reg r8
```

Comment:
We should not be using push_nonvol_reg for r8 and r9 since they are volatile. Beside, I don't think we can use r8 here since it stores the pThread parameter.